lib/uplink/buckets.go

```diff
@@ -60,3 +63,84 @@ func (s *Session) ListBuckets(ctx context.Context, opts storj.BucketListOptions)
 	storj.BucketList, error) {
 	return storj.BucketList{}, nil
 }
+
+// DeleteBucket deletes a bucket and returns an error if there was a problem
+func (client *Client) DeleteBucket(ctx context.Context, bucket string) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	list, err := client.metainfo.ListObjects(ctx, bucket, storj.ListOptions{Direction: storj.After, Recursive: true, Limit: 1})
+	if err != nil {
+		return convertError(err, bucket, "")
+	}
+
+	if len(list.Items) > 0 {
+		return minio.BucketNotEmpty{Bucket: bucket}
+	}
+
+	err = client.metainfo.DeleteBucket(ctx, bucket)
+
+	return convertError(err, bucket, "")
+}
+
+// GetBucketInfo gets the bucket information and returns the info and an error type
+func (client *Client) GetBucketInfo(ctx context.Context, bucket string) (bucketInfo minio.BucketInfo, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	info, err := client.metainfo.GetBucket(ctx, bucket)
+
+	if err != nil {
+		return minio.BucketInfo{}, convertError(err, bucket, "")
+	}
+
+	return minio.BucketInfo{Name: info.Name, Created: info.Created}, nil
+}
+
+// ListBuckets returns a list of buckets in the store
+func (client *Client) ListBuckets(ctx context.Context) (bucketItems []minio.BucketInfo, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	startAfter := ""
+
+	for {
+		list, err := client.metainfo.ListBuckets(ctx, storj.BucketListOptions{Direction: storj.After, Cursor: startAfter})
+		if err != nil {
+			return nil, err
+		}
+
+		for _, item := range list.Items {
+			bucketItems = append(bucketItems, minio.BucketInfo{Name: item.Name, Created: item.Created})
+		}
+
+		if !list.More {
+			break
+		}
+
+		startAfter = list.Items[len(list.Items)-1].Name
+	}
+
+	return bucketItems, err
+}
+
+// MakeBucketWithLocation makes a bucket with a location and returns an error if there were any problems.
+func (client *Client) MakeBucketWithLocation(ctx context.Context, bucket string, location string) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	// TODO: This current strategy of calling bs.Get
```

Comment:
Yes, and a similar problem occurs with DeleteBucket: there is a race between checking if the bucket is empty and then deleting it. And from my trying to follow it down, it looks to me like DeleteBucket itself doesn't check for that sort of thing; it just says "ok" and removes the bucket data whether there are entries there or not. The "don't Delete non-empty buckets" and "don't Put buckets that already exist" rules should really be enforced by the pointerdb if we care about them at all. And then in that case, there would be no need to do these racy checks; you'd just call DeleteBucket/CreateBucket/etc and report whatever errors came back.

But fixing that would be way out of scope for this ticket, and as you say, this is the way it's already done in the gateway and uplink command-line tool, so this is good for now :)